File: LeetCode/322.Coin_Change/322.Coin_Change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        dp[0] = 0;

        for (int i = 1; i < dp.length; i++) {
            int res = -1;
            for (int c : coins) {
                if (i >= c && dp[i - c] != -1) {
                    if (res == -1) {
                        res = dp[i - c] + 1;
                    } else {
                        res = Math.min(res, dp[i - c] + 1);
                    }
                }
            }
            dp[i] = res;
        }
        return dp[dp.length - 1];
    }
}

File: LeetCode/322.Coin_Change/322.Coin_Change.swift
class Solution {
    func coinChange(_ coins: [Int], _ amount: Int) -> Int {
        var dp = Array(repeating: amount + 1, count: amount + 1)

        // init dp
        dp[0] = 0

        for coin in coins {
            if amount < coin { continue }
            for i in coin ... amount {
                dp[i] = min(dp[i - coin] + 1, dp[i])
            }
        }

        return dp[amount] == amount + 1 ? -1 : dp[amount]
    }
}
